android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/SessionPrioritizationMiddleware.kt

```diff
@@ -60,21 +98,34 @@ class SessionPrioritizationMiddleware : Middleware<BrowserState, BrowserAction>
         }
     }
 
-    private fun updatePriorityIfNeeded(state: BrowserState) {
+    private fun updatePriorityIfNeeded(state: BrowserState) = mainScope.launch {
         val currentSelectedTab = state.selectedTabId?.let { state.findTab(it) }
         val previousSelectedTab = state.findTab(previousHighestPriorityTabId)
         val currentEngineSession: EngineSession? = currentSelectedTab?.engineState?.engineSession
 
-        // We need to make sure, we alter the previousHighestPriorityTabId, after the session is linked.
+        // We need to make sure we alter the previousHighestPriorityTabId, after the session is linked.
         // So we update the priority on the engine session, as we could get actions where the tab
         // is selected but not linked yet, causing out sync issues,
         // when previousHighestPriorityTabId didn't call updateSessionPriority()
         if (currentEngineSession != null) {
-            previousSelectedTab?.engineState?.engineSession?.updateSessionPriority(DEFAULT)
+            mainScope.launch {
+                // check for existing form data here and only set DEFAULT if there is
+                previousSelectedTab?.engineState?.engineSession?.checkForFormData()
+            }
+
             currentEngineSession.updateSessionPriority(HIGH)
             logger.info("Update the currentSelectedTab ${currentSelectedTab.id} priority to ${HIGH.name}")
-            logger.info("Update the previousSelectedTab ${previousSelectedTab?.id} priority to ${DEFAULT.name}")
             previousHighestPriorityTabId = currentSelectedTab.id
         }
     }
+
+    private fun clearHighPriority(context: MiddlewareContext<BrowserState, BrowserAction>, tabId: String) {
```

Comment:
We should probably add some tests for this to know that we are dispatching the action to clear high priority.

There is a `CaptureActionsMiddleware` you can add to your test `BrowserStore` to catch the actions that go through the store and make assertions with it.

Reply:
So when I was adding this test, `CaptureActionsMiddleware` errored out with `no action found` even though I can see printouts from the clear high priority action. I did notice it was able to capture the action when it passes through the `ContentStateReducer` instead of throwing the exception `throw IllegalStateException("You need to add SessionPrioritizationMiddleware. ($action)")`. I observed the same thing with other actions that also throw exceptions `RemoveThumbnailAction, UpdateThumbnailAction`. I'm wondering if this is a possible limitation to `CaptureActionsMiddleware`.

Let me know if it's not clear, we can talk about it

Reply:
Are you adding the `CaptureActionsMiddleware` as the last middleware in the list?

When I tried to replicate what you described above with the `ThumbnailsMiddlewareTest`, I can comment out the `UpdateThumbnailAction` action in the test to see the middleware fail correctly.

<details>

<summary>Stack trace</summary>

```
Action of type class mozilla.components.browser.state.action.ContentAction$UpdateThumbnailAction was dispatched: UpdateThumbnailAction(sessionId=test-tab1, thumbnail=Mock for Bitmap, hashCode: 1365863352)
java.lang.AssertionError: Action of type class mozilla.components.browser.state.action.ContentAction$UpdateThumbnailAction was dispatched: UpdateThumbnailAction(sessionId=test-tab1, thumbnail=Mock for Bitmap, hashCode: 1365863352)
	at mozilla.components.support.test.middleware.CaptureActionsMiddleware.assertNotDispatched(CaptureActionsMiddleware.kt:75)
	at mozilla.components.browser.thumbnails.ThumbnailsMiddlewareTest.thumbnail actions are the only ones consumed by the middleware(ThumbnailsMiddlewareTest.kt:171)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
```

</details>

Reply:
Ya I added it as the last middleware.

If you still have it, would you be able to paste the modified test?  When I have just this in the test, it should work but won't capture the action.

```
store.dispatch(ContentAction.UpdateThumbnailAction("test-tab1", mock())).joinBlocking()
capture.assertLastAction(ContentAction.UpdateThumbnailAction::class) {}
```

Reply:
Talked to Jon and resolved this by specifying the order of middleware in the unit test as `middleware = listOf(capture, middleware)`. This allows the capture middleware to be invoked first. That way the action is added to the list of `capturedActions` before it reaches the SessionPrioritizationMiddleware and is returned without reaching the `ContentStateReducer`.